### ISSUE TITLE
[Request] Broadcast Message to Multiple Channels

### ISSUE THREAD

**Issue by gravypod** [sustrik/libdill#52]
Hello, and thanks for reading this (again). The semester has been really busy and ironically due to my studies I have not been able to... well.. continue my studies. I've got a bit of time coming up after finals and the project that initially prompted this request has started to pear it's head over the horizon so I think I'll pose this again.

My initial request is as follows. I've taken the liberty to adjust the language used to be more suitable for use with this project:

This is more of a feature request to see if such a thing is possible. 

I was thinking of writing an IRC server or some such system with libdill (changed from libmill) and I was looking through the docs. It doesn't seem like this sort of system is supported without managing an array of channels (one for each listener).

This feature was also brought up in a mailing list that I found when googling: http://www.freelists.org/post/libmill/broadcast-effect

And the answer that was proposed seemed more like an anti-pattern in this case: http://www.freelists.org/post/libmill/broadcast-effect,1 

Is it possible to support 

```
coroutine void on_c(int channel)
{
  char *message;
  int err = chrecv(channel, &message, sizeof(char*), -1);
  if (err) message = "Error handling message";
  printf("%s\n", message);
}


void main()
{
  char *message = "Hello World";
  // Create a broadcast channel 
  int ch = bchannel(sizeof(int), 0);
  // Pass this to many coroutines
  go(on_c(ch));
  go(on_c(ch));
  go(on_c(ch));
  // Broadcast this to listeners. 
  chbroadcast(ch, &message, sizeof(char*), -1);
}

```
Where the output is:

> Hello World
> Hello World
> Hello World

This would, in my opinion, create some interesting design patterns for message/event driven systems in C.

Do you think something like this would be detrimental to this library? How would someone go about implementing this into the code base if you aren't interested in doing this? Would a feature like this be merged/mainlined or are you uninterested in supporting a feature like this within this platform?

I'm very excited about the prospects of being able to design a structured concurrency system that can take advantage of an API like this.

**Comment by raedwulf** [comment]
We're looking to change the behaviour of channels at the moment but I think this statement still holds.  If you want to make a broadcast channel, you need to consider how the messages are consumed from the channel.

Any solution involving a single buffer requires all messages waiting on top of the buffer until all other readers have completed.  Any solution involving multiple buffers (one for each reader) can be implemented using multiple channels.

If you want to implement broadcast i was recommend writing a wrapper object that uses multiple channels under the hood.

**Comment by gravypod** [comment]
It's there a way to check if a channel is open still? That's the only way I
can see a high performance solution being possible.

**Comment by raedwulf** [comment]
I'm not sure what you're getting at.  Existing readers and writers blocked on a channel do receive EPIPE error immediately when you send `chdone`.

It is better to use `chdone` on the channel as opposed to `hclose` it straight away as a coroutine may be blocked on another event when the channel is closed and that is not easily detectable.
`hclose` the channel after the coroutines using them complete.

**Comment by gravypod** [comment]
I think I'd need some way to see if a channel is still opened to implement
a broadcast channel.

The broadcast channel needs to have a collection of channels to forward to.
This needs to have old channels either removed or ignored on sending.
Ignore on sending will likely be the easiest method of doing this.

**Comment by raedwulf** [comment]
`chdone` from within the coroutine can indicate to the broadcast channel that the channel is no longer used and `EPIPE` error will result from using it in the broadcast channel.
The broadcast channel should own the individual channels to make it easy to manage, so it is responsible for `hclose` individual channel handles afterwards.

**Comment by gravypod** [comment]
Sorry I've taken a bit to get back to this. I've implemented a basic broadcast channel and I've just gotten it to a semi-stable state. I'd like it if you could give me any comments on the implementation I have here. Any information as to what I should consider changing would be great. It's taken me a bit to get into finishing this because I had some completely incorrect perceptions of the functionality of some portions of this library. 

You can find my implementation here: 
https://gist.github.com/gravypod/7e0cda2f11ef719d39a515d6986ba892

I've taken your advice @raedwulf and I use EPIPE as a signal for removing a channel from my list of receivers. 

I've went with a few design choices that I cannot judge by myself as I have a limited understanding of these libraries. 

1. I allocate a buffer in `coroutine void forward` to buffer the object being recived before it is sent out.
2. I've implemented the function that loops over all of the channels and actually forwards the messages as a coroutine.
3. Error codes that are not EPIPE are not handled at all. 
4. Using blisten to create a child channel for a reciver. I have no idea if this is what I should be doing. 
5. Can coroutines run on different processes? Do I need to use some form of reentrant locking for synchronization on mutation operations?

I'm also considering adding another primative. I'm thinking of calling it a `package`. The idea is that you can give a package a pointer and a destructor method. This can be passed to N functions. After N functions take the value the `package` calls the destructor. 

So for instance:

```
package_t *p = NULL;
int result = ptake(ch);
char *message = (char*) p->value;
printf("%s\n", message);
prelease(p);
```

If this is broadcasted to 10 channels, then after you see 10 releases the destructor given to it will be run on `p->value` and the package object itself will be free'd. This way if you have a complext object that is constructed and must be processed N ways you can be sure it will be deleted after the computation completes. This is basically a poor-man's reference-counting GC. 

For now I'm going to go study for my exams. Any comments or suggestions are appreciated.

**Comment by sustrik** [comment]
You don't need that much complexity IMO. Just keep a list of channels and on send() iterate over them and chsend(msg,size,-1) the message to each of them. Done.

That will of course block if there is a non-responsive receiver, but just that's the nature of reliable broadcast.

If, OTOH, you won't non-reliable non-blocking broadcast use chsend(msg,size,0). That will send the message if receiver is responsive or skip it if it's not.

**Comment by raedwulf** [comment]
Some additional points:

- As Martin said, broadcast can only be reliable or unreliable; reliable will always cause blocking and non-reliable will drop messages for some receiver.
- Abstraction for a broadcast channel object is probably unnecessary - just have a function that broadcasts over a list of channels and maintain that channel list in a high-level object.  Otherwise you're adding complexity where there is none required.  If you start finding yourself having to write the channel-list construction, destruction more than once in different modules then do consider abstracting a broadcast channel in your application (that is a valid use-case).
- Having said that, this is a good exercise for learning libdill.

> Can coroutines run on different processes? Do I need to use some form of reentrant locking for synchronization on mutation operations?

In the current version, libdill contexts are unique per thread which gets automatically created when libdill functions are called in a separate thread.  However, coroutines cannot be migrated and you cannot share libdill-created handles.  The reason for this is performance and ease of use.  The primary function of libdill is to have very fast user-threads and allowing migration and thread-safety across threads will go against this goal (thread synchronisation within a single context is very difficult and slow).

The shared library build of libdill compromises slightly performance-wise and that was unavoidable to get thread-safe contexts.  However, if you use the static build, the performance of thread-safe libdill matches that of the single-threaded version (< 40ns coroutine creation and <15 ns switches between contexts).

To communicate between different libdill contexts I suggest using two unix pipes or a single unix socket for synchronisation and communication between coroutines across processes.

> I was thinking of writing an IRC server or some such system with libdill (changed from libmill) and I was looking through the docs. It doesn't seem like this sort of system is supported without managing an array of channels (one for each listener).

That sounds great! Have a look at https://github.com/sustrik/dsock and consider sending a pull request for the IRC protocol.

**Comment by raedwulf** [comment]
Also when considering your design have a look at this resources if you have not already done so:

- http://libdill.org/structured-concurrency.html
- http://250bpm.com/blog:73 (as an update to this, we've modified `choose` behaviour to be deterministic and hence fill a use-case not covered by this behaviour).  This is explained further in this quote.

Martin Sustrik:
> The reasoning is as follows: In golang, clause is chosen at random to ensure fairness among channels and prevent any of them from being starved. While that makes sense if all the channels are of the same type (i.e. they are semantically equivalent) libdill's design is based on assumption that all semantically equivalent messages are passed via a single channel (possibly with multiple senders). Therefore, if choose is used at all, the channels in question are likely to differ semantically. And if that's the case, prioritization based on type is preferable to using one type at random.

Since you started implementing it based on buffered channels, we've simplified it to unbuffered channels.  My quote:
> So another way of thinking about it is that the [channel] buffer (copying) can be described as a lightweight process (the memcpy essentially). The only difference is that the buffer stores the results somewhere intermediate and a second process (the receiver) starts reading it afterwards.

Issue #51 is our main discussion on this where the quotes come from.

**Comment by gravypod** [comment]
> You don't need that much complexity IMO. Just keep a list of channels and on send() iterate over them and chsend(msg,size,-1) the message to each of them. Done.
> 
> That will of course block if there is a non-responsive receiver, but just that's the nature of reliable broadcast.
> 
> If, OTOH, you won't non-reliable non-blocking broadcast use chsend(msg,size,0). That will send the message if the receiver is responsive or skip over it if it's not.

This is organized in this way for multiple reasons.

1. The channel_collection_t was made because I wanted to make sure that the hot loop, looping over everything, would be able to hit the cache as much as possible. That is why I do my best to linerally address all of the channels. All of the code in that file is for printing out the state,  creation, adding, and removing the item as well as recollecting the entire object. I don't see a better way to do do this. 
2. I feel that this is a basic operation that should be in this type of library. It's something one would expect to find. If there's One to One, Many to One, why isn't there One to Many?
3. The extra coroutines are there to make sure we don't block real work. Those, in my understanding, will only be run once a thread does something blocking. In this case we want to waste clock cycles spining over an array of ints. In most other cases we don't. For instance in the middle of large buisness logic that want's to send out an event but still needs to finish the sutff after it.

>As Martin said, broadcast can only be reliable or unreliable; reliable will always cause blocking and non-reliable will drop messages for some receiver.

I've taken a lot of care to hopefully avoid that problem. There is, in my opinion, excessive buffering going on in the bchannel. Each listener has a buffer the same size as the master channel. This is to make sure that the master channel is to block before all of the children do. This way anything hung should be allowed to finish. 

>Abstraction for a broadcast channel object is probably unnecessary - just have a function that broadcasts over a list of channels and maintain that channel list in a high-level object. Otherwise you're adding complexity where there is none required. If you start finding yourself having to write the channel-list construction, destruction more than once in different modules then do consider abstracting a broadcast channel in your application (that is a valid use-case).

> Since you started implementing it based on buffered channels, we've simplified it to unbuffered channels. My quote: ...

I'll read through that in a bit. 

I don't think that's the solution here expecially if you're sending messages from multiple places. I think I've done a good job at taming the complexity here.  The idea is to have a broadcast channel as a drop in library and I don't think creating and maintaining a list of channels in your buisness logic is such a great idea as there are evidently a lot of ifs, ands, butts, and coconuts that go along with this. 

I spent an embarisngly long amount of time finding a bug in this: https://gist.github.com/gravypod/7e0cda2f11ef719d39a515d6986ba892#file-bchannel-c-L17

You need to iterate backwards in case there is a remove, that is evident from the start, but what isn't evident is that if you're using a size_t iterating from `for (size_t i = N - 1; i >= 0; i--)` then you're going to never leave that loop. Things like that make this problem unfun to do in the center of something else. It's the old "I'm writing X not an message passing system!". Currently I'm writing the message passing system so I can write the X.

**Comment by gravypod** [comment]
The back end might seem complex so I've implemented a simple test script: https://gist.github.com/gravypod/3506ec5524b3be6963d280f9a144035e

It should demonstrate that this just looks just like normal channels despite a few changes here and there. I intend to implement a `bdone(b)` instead of `chdone(b->submission)`. I think this system looks as simple as it can be. I'm not sure where I can trim the fat here if I'm attempting to make this a general library.

**Comment by raedwulf** [comment]
> The extra coroutines are there to make sure we don't block real work. Those, in my understanding, will only be run once a thread does something blocking. In this case we want to waste clock cycles spining over an array of ints. In most other cases we don't. For instance in the middle of large buisness logic that want's to send out an event but still needs to finish the sutff after it.

Yes that's fine

>  The idea is to have a broadcast channel as a drop in library and I don't think creating and maintaining a list of channels in your buisness logic is such a great idea as there are evidently a lot of ifs, ands, butts, and coconuts that go along with this.

Sorry, I was initially confused as I assumed this was directed for the IRC server project as opposed to a more general approach.  We changed channels to become unbuffered (and only unbuffered) because it makes it more of a primitive building block so that you can implement different types of channels more easily without being enforced to use any one type of buffer, or, in your case, one-to-many communication and buffering.  Creating a library for channel-based message-passing is a worthwhile endeavour.

> The channel_collection_t was made because I wanted to make sure that the hot loop, looping over everything, would be able to hit the cache as much as possible. That is why I do my best to linerally address all of the channels. All of the code in that file is for printing out the state, creation, adding, and removing the item as well as recollecting the entire object. I don't see a better way to do do this.

Unfortunately, the channel structures are accessed within libdill and although it's likely true that successively malloc'd channels will be addressed nearby, there's certainly no guarantees that this is the behaviour.  I'm looking to improving my memory pool implementation so hopefully this will likely be the case.

I feel the collection of channels object might be unnecessary tied to the bchannel implementation and it's more C-idiomatic to pass an array of int handles with a specified length. 

@sustrik What do you think?

**Comment by gravypod** [comment]
>  We changed channels to become unbuffered (and only unbuffered) because it makes it more of a primitive building block so that you can implement different types of channels more easily without being enforced to use any one type of buffer, or, in your case, one-to-many communication and buffering

Yea perfect, that's ecacly what I need to make this (more) efficient. I've got the stuff under the hood to manage my own buffering for this application.

> Unfortunately, the channel structures are accessed within libdill and although it's likely true that successively malloc'd channels will be addressed nearby, there's certainly no guarantees that this is the behaviour. I'm looking to improving my memory pool implementation so hopefully this will likely be the case.

That's unfortunate but I still think a serial loop like what I'm doing will be faster then a linked-list structure. That's what I started with but found this to be simpler and probably faster. 

> Is the reason for the collection to allow the bchannel to expand the number of listeners?

Definetly. It's also made so that we can dynamically remove channels that have died. The reason for this was to allow N recivers and for dynamic removal and allocation of more listeners. It's slightly complicated as it automatically frees up some memory when we have some available and for a penalty of 1 malloc and setting 3 variables. There is an aditional malloc after every 9 adds. So not too bad right now since that's more of a setup thing.

**Comment by raedwulf** [comment]
Btw, keep in mind if you're actually sending a message to each client (your IRC app), a syscall is going to be far more expensive than any micro-optimisation occuring here.

**Comment by sustrik** [comment]
As for the allocation stuff, I've added chmake_s function that lets you create a channel in memory that you supply.

**Comment by raedwulf** [comment]
Mmm slightly related,  naming wise should we have `go_s` as well? Or should we avoid C11 annex k safe function names entirely to avoid confusion.

**Comment by sustrik** [comment]
Oops, I haven't considered C11. Any better naming for functions that take storage parameter?

**Comment by raedwulf** [comment]
Technically, C doesn't normally have `*make` style functions, but since `printf`-style functions have an implicit stream, we could use that idea that `chmake` does an implicit buffer.

`sprintf` writes to a string - so C-functions prefer prefixing buffer/types in front of the function name although that does conflict with consistently prefixing all channel related names with `ch`.  So maybe we follow the `FILE` stream convention, `fgetc`, `fgets` where it suffixes the buffer/type at the end.

So options could be (using convention that the buffer is typically denoted as `mem` or, although not used anywhere, `m` could be a shorter form).

- `memchmake` or `chmakemem` or `chmemmake` - prefix (match stdlib.h) or suffix the function (prevent inconsistent prefix) with a buffer or embed within the function name.  Similarly, shortforms: `mchmake`, `chmakem`, `chmmake`.  This implies that there's a `len` parameter though (so maybe `go_stack` would be a better candidate for this name change).
- `chmakes` - remove the underscore to match file stream type functions (but technically not a string so possibly misleading still)... Maybe use `chmakestor` or `chstormake` to denote `chstorage`?
Or replace `stor` with `stg`.

The `chstorage` approach makes sense from an ABI perspective although as far as I'm aware, the only C-library function that takes the same approach is `setjmp`, `longjmp` and `jmp_buf`.  As `chstorage` is a type, it seems that the C standard convention often uses underscores and the only example uses `_buf` in that instance.. so maybe `ch_buf`? Although this might lead to confusion as channels are unbuffered in a different respect so maybe `ch_storage` or `ch_stor` or `ch_stg`.

Another side note: maybe we could generate the minimum buffer size required for `chstorage` using autotools rather than being extra conservative? This would break binary compatibility between channel ABI changes... but we'd still have source compatibility.

**Comment by sustrik** [comment]
The idiomatic way to do this is to use the name of what you are creating as the name of the function (socket, pipe etc.) However, using 'channel' and 'handle' is asking for name clashes.

I've tried 'hcreate' but that clashes with POSIX function of the same name.

Thus make.

As for the suffix, 'm' or 'mem' sounds good, but I would prefer them to be separated by underscore so that it's clear that they are just variants of the original function. go_m? chmake_m?

As for the storage type, there's also the _t convention. Not sure what's the best approach is. I have to think about it a bit more.

**Comment by gravypod** [comment]
> As for the allocation stuff, I've added chmake_s function that lets you create a channel in memory that you supply.

Is it 1 chstorage to 1 channel or can I share N channels being backed by the cache of 1 chstorage object?

**Comment by raedwulf** [comment]
> As for the suffix, 'm' or 'mem' sounds good, but I would prefer them to be separated by underscore so that it's clear that they are just variants of the original function. go_m? chmake_m?

I'd probably go for the longer form if we add the underscore, `go_mem` and `chmake_mem`.  The `_t` convention could work too although I think it typically is used for typedef'd structs?

**Comment by raedwulf** [comment]
> Is it 1 chstorage to 1 channel or can I share N channels being backed by the cache of 1 chstorage object?

It's 1:1

**Comment by sustrik** [comment]
I went with the _mem convention. Merged into the mainline now.
